yarn-project/simulator/src/private/providers/circuit_recording/simulation_provider_recorder_wrapper.ts

```diff
@@ -0,0 +1,88 @@
+import { createLogger } from '@aztec/foundation/log';
+import type { ForeignCallHandler } from '@aztec/noir-protocol-circuits-types/types';
+import type { FunctionArtifactWithContractName } from '@aztec/stdlib/abi';
+import type { NoirCompiledCircuitWithName } from '@aztec/stdlib/noir';
+
+import type { ACIRCallback, ACIRExecutionResult } from '../../acvm/acvm.js';
+import type { ACVMWitness } from '../../acvm/acvm_types.js';
+import type { SimulationProvider } from '../simulation_provider.js';
+import { CircuitRecorder } from './circuit_recorder.js';
+
+/**
+ * Takes a simulation provider and wraps it in a circuit recorder. See CircuitRecorder for more details on how circuit
+ * recording works.
+ */
+export class SimulationProviderRecorderWrapper implements SimulationProvider {
+  constructor(private simulator: SimulationProvider, private log = createLogger('simulator-recorder')) {}
+
+  async executeProtocolCircuit(
```

Comment:
I'd merge the impl of these two - the only difference is the simulator fn. You can abstract that into `fn simulate(callback) => Promise<void>`.